File: backend/app/ai/prompts.py
"""Prompt factories — mirrors src/lib/ai/prompts.ts exactly."""

LANGUAGE_INSTRUCTIONS = {
    "en": "Respond entirely in English.",
    "ru": "Отвечай полностью на русском языке.",
    "kk": "Толығымен қазақ тілінде жауап бер.",
}


def language_instruction(lang: str) -> str:
    return LANGUAGE_INSTRUCTIONS.get(lang, LANGUAGE_INSTRUCTIONS["en"])


# ── Lesson ────────────────────────────────────────────────────────────────────
def lesson_system() -> str:
    return (
        "You are an expert curriculum designer and master teacher. "
        "You create detailed, practical, and engaging lesson plans. "
        "Always respond in well-structured Markdown."
    )


def lesson_user(subject: str, grade: str, topic: str, duration: int, lang: str, notes: str | None) -> str:
    extra = f"\n\nAdditional notes: {notes}" if notes else ""
    return f"""\
Create a complete lesson plan with these exact sections:

## Learning Objectives
## Materials Needed
## Lesson Flow
(A markdown table with columns: Phase | Duration | Teacher Activity | Student Activity)
## Core Explanation
## Practice Activities
## Differentiation Strategies
## Homework Assignment
## Discussion Questions
## Exit Ticket

Details:
- Subject: {subject}
- Grade/Level: {grade}
- Topic: {topic}
- Duration: {duration} minutes
- {language_instruction(lang)}{extra}

Be specific, practical, and classroom-ready."""


# ── Quiz ─────────────────────────────────────────────────────────────────────
def quiz_system() -> str:
    return (
        "You are an expert educator who creates high-quality assessments. "
        "Respond with ONLY valid JSON — no prose, no code fences."
    )


def quiz_user(subject: str, grade: str, topic: str, num: int, difficulty: str, types: list[str], lang: str) -> str:
    type_str = ", ".join(types)
    return f"""\
Create a {difficulty} quiz on "{topic}" for {subject}, {grade}.
Generate exactly {num} questions using these types (cycle if needed): {type_str}.

Return ONLY this JSON structure:
{{
  "title": "Quiz title",
  "questions": [
    {{
      "id": 1,
      "type": "multiple_choice",
      "question": "...",
      "options": ["A", "B", "C", "D"],
      "answer": "A",
      "explanation": "..."
    }},
    {{
      "id": 2,
      "type": "true_false",
      "question": "...",
      "options": ["True", "False"],
      "answer": "True",
      "explanation": "..."
    }},
    {{
      "id": 3,
      "type": "open",
      "question": "...",
      "answer": "Model answer...",
      "explanation": "..."
    }}
  ]
}}

{language_instruction(lang)}
Return ONLY the JSON object, nothing else."""


# ── Grade ─────────────────────────────────────────────────────────────────────
def grade_system() -> str:
    return (
        "You are a fair, constructive teacher who grades student work. "
        "Respond with ONLY valid JSON — no prose, no code fences."
    )


def grade_user(subject: str, grade: str, question: str, answer: str, rubric: str | None, lang: str) -> str:
    rubric_line = f"\nRubric: {rubric}" if rubric else ""
    return f"""\
Grade this student answer.

Subject: {subject}
Grade level: {grade}
Question: {question}
Student's answer: {answer}{rubric_line}

Return ONLY this JSON:
{{
  "score": 0-100,
  "summary": "One sentence overall assessment.",
  "strengths": ["strength 1", "strength 2"],
  "mistakes": ["mistake 1", "mistake 2"],
  "recommendations": ["recommendation 1", "recommendation 2"]
}}

{language_instruction(lang)}
Return ONLY the JSON object, nothing else."""


# ── Assistant ─────────────────────────────────────────────────────────────────
def assistant_system(lang: str) -> str:
    return (
        f"You are Teacher Copilot, an expert teaching assistant. "
        f"You give practical, classroom-ready advice to teachers. "
        f"Be concise, actionable, and supportive. {language_instruction(lang)}"
    )

File: frontend/src/lib/ai/prompts.ts
import type { GradeInput, LessonInput, QuizInput, Language } from "../types";
import { QUESTION_TYPE_LABELS } from "../types";
import { languageInstruction } from "../utils";

export interface PromptPair {
  system: string;
  user: string;
}

// ── Lesson Planner ──────────────────────────────────────────
export function lessonPrompt(input: LessonInput): PromptPair {
  const system = [
    "You are Teacher Copilot — an expert instructional designer and pedagogy coach.",
    "You help teachers create clear, engaging, age-appropriate lesson plans grounded in good pedagogy (clear objectives, active learning, formative assessment, differentiation).",
    "Always respond in clean, well-structured GitHub-flavored Markdown using `##` section headings, bullet lists, and tables where helpful. Do not include a top-level `#` title.",
  ].join(" ");

  const user = [
    `Create a complete, ready-to-teach lesson plan.`,
    ``,
    `- Subject: ${input.subject}`,
    `- Grade / level: ${input.grade}`,
    `- Topic: ${input.topic}`,
    `- Total duration: ${input.durationMinutes} minutes`,
    input.notes ? `- Teacher notes: ${input.notes}` : ``,
    ``,
    `Include these sections:`,
    `1. **Learning Objectives** — 3–4 measurable objectives ("Students will be able to…").`,
    `2. **Materials & Preparation** — what the teacher needs.`,
    `3. **Lesson Flow** — a Markdown table with columns: Phase | Time | What the teacher does | What students do. The times must add up to ${input.durationMinutes} minutes.`,
    `4. **Core Explanation** — a concise, correct explanation of the topic written at the right level for ${input.grade}.`,
    `5. **Guided & Independent Practice** — concrete activities.`,
    `6. **Differentiation** — support for struggling learners and extension for advanced learners.`,
    `7. **Homework** — a short, meaningful assignment.`,
    `8. **Discussion Questions** — 4–5 open questions to spark thinking.`,
    `9. **Exit Ticket** — one quick formative check.`,
    ``,
    languageInstruction(input.language),
  ]
    .filter(Boolean)
    .join("\n");

  return { system, user };
}

// ── Quiz Generator ──────────────────────────────────────────
export function quizPrompt(input: QuizInput): PromptPair {
  const allowed = input.types.map((t) => QUESTION_TYPE_LABELS[t]).join(", ");

  const system = [
    "You are Teacher Copilot — an expert assessment designer.",
    "You write accurate, fair, curriculum-aligned quiz questions.",
    "You output ONLY valid minified JSON that matches the requested schema. No prose, no Markdown, no code fences.",
  ].join(" ");

  const user = [
    `Generate a quiz as a single JSON object.`,
    ``,
    `Parameters:`,
    `- Subject: ${input.subject}`,
    `- Grade / level: ${input.grade}`,
    `- Topic: ${input.topic}`,
    `- Number of questions: ${input.numQuestions}`,
    `- Difficulty: ${input.difficulty}`,
    `- Allowed question types: ${allowed}`,
    ``,
    `JSON schema (keys MUST stay in English):`,
    `{`,
    `  "title": string,`,
    `  "questions": [`,
    `    {`,
    `      "id": number,                       // 1-based`,
    `      "type": "multiple_choice" | "true_false" | "open",`,
    `      "question": string,`,
    `      "options": string[],                // 4 items for multiple_choice; ["True","False"] for true_false; omit for open`,
    `      "answer": string,                   // exact correct option, or a model answer for open questions`,
    `      "explanation": string               // why the answer is correct`,
    `    }`,
    `  ]`,
    `}`,
    ``,
    `Only use question types from the allowed list. Vary them across the quiz. Make questions genuinely about "${input.topic}".`,
    `The visible text (question, options, answer, explanation) must follow this instruction: ${languageInstruction(
      input.language,
    )}`,
  ].join("\n");

  return { system, user };
}

// ── Assignment Grader ───────────────────────────────────────
export function gradePrompt(input: GradeInput): PromptPair {
  const system = [
    "You are Teacher Copilot — a fair, encouraging, and rigorous grading assistant.",
    "You evaluate student work constructively: you recognise what is correct, identify mistakes precisely, and give actionable next steps.",
    "You output ONLY valid minified JSON that matches the requested schema. No prose, no Markdown, no code fences.",
  ].join(" ");

  const user = [
    `Grade the following student answer.`,
    ``,
    `- Subject: ${input.subject}`,
    `- Grade / level: ${input.grade}`,
    `- Question / task: ${input.question}`,
    input.rubric ? `- Rubric / criteria: ${input.rubric}` : ``,
    ``,
    `Student answer:`,
    `"""`,
    input.studentAnswer,
    `"""`,
    ``,
    `Return a single JSON object (keys MUST stay in English):`,
    `{`,
    `  "score": number,                 // 0-100`,
    `  "summary": string,               // 1-2 sentence overall judgement`,
    `  "strengths": string[],           // what the student did well`,
    `  "mistakes": string[],            // specific errors or gaps`,
    `  "recommendations": string[]      // concrete next steps / topics to review`,
    `}`,
    ``,
    `Be specific and reference the student's actual answer. The visible text must follow: ${languageInstruction(
      input.language,
    )}`,
  ]
    .filter(Boolean)
    .join("\n");

  return { system, user };
}

// ── Teacher Assistant (chat) ────────────────────────────────
export function assistantSystem(language: Language): string {
  return [
    "You are Teacher Copilot — a knowledgeable, practical teaching assistant for K-12 and secondary teachers.",
    "You give classroom-ready advice: lesson ideas, explanations pitched to a grade level, activities, behaviour strategies, and assessment help.",
    "Be concise and actionable. Use Markdown (short paragraphs, bullet lists, bold key terms). When a teacher mentions a grade or age, tailor your answer to it.",
    languageInstruction(language),
  ].join(" ");
}
